cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java

```diff
@@ -434,10 +455,16 @@ public boolean hasOngoingExecution() {
     private ExecutorState.State _state;
     private Set<Integer> _recentlyDemotedBrokers;
     private Set<Integer> _recentlyRemovedBrokers;
+    private final long _executionStartMs;
+    private Throwable _executionException;
+    private UserTaskManager.UserTaskInfo _userTask;
 
     ProposalExecutionRunnable(LoadMonitor loadMonitor, Collection<Integer> demotedBrokers, Collection<Integer> removedBrokers) {
       _loadMonitor = loadMonitor;
       _state = ExecutorState.State.NO_TASK_IN_PROGRESS;
+      _executionStartMs = _time.milliseconds();
+      _executionException = null;
+      _userTask = _userTaskManager == null ? null : _userTaskManager.getUserTaskById(_uuid);
```

Comment:
looks like this does not solve the issue since you are just get the reference not clone the `UserTaskInfo`, in this way we still need fix patch for 598.

Also should we panic(throw exception) here if the _userTaskManager is null(which means we forget to pass in reference to Executor).

Reply:
The reference is not garbage collected since we still have a reference here even if we remove it in `UserTaskManager`. 
Panic'ing may be too harsh, since `ProposalExecutionRunnable` is also used in unit tests, if `UserTaskManager` isn't set, we just won't have complete info.

Reply:
yes, you are right. Another question is how do you guarantee visibility across the thread, I think if you want to make sure `_userTask` is in-sync, you need `volatile` or `AtomicReference`.

I still feel panic is the right thing to do, in unit test you can pass in a mock object. I think in your current implementation, `UserTaskManager` is a necessary thing since you try to get `UserTaskInfo` from it and whole purpose of this patch is return these information to user.

Reply:
I'll make it AtomicReference.
When you say panic, do you mean throwing an exception? I think even if we don't have user task information, it's still better to send the email out instead of panic out, since if we send nothing, the user will not know execution has finished, resulting in unnecessary waiting, whereas if we send out the email with UUID, the user can lookup the info, either via /user_tasks (within retention), or logs.

Reply:
I am very confused.

@alexwang08 When would you expect the `_userTaskManager` to be `null`? If it is just for unit tests, why don't we use mocks in the unit tests? If it indicates that your code logic is wrong, it should fail -- would you clarify under which scenario this is true: `if we don't have user task information, it's still better to send the email out instead of panic out`

@kidkun has this issue been resolved? If not, why did we mark the patch as approved?

Reply:
@efeg had a discussion offline, Alex's argument is that in that case we send out the email without information in `UserTaskInfo`.

Reply:
@efeg `_userTaskManager` is set in `KafkaCruiseControlServlet's` constructor. My code is right. `_userTaskManager` helps to tell if the execution is started by user, what operation it is, and request parameters etc. These information does not affect correctness, so I didn't set `_userTaskManager` in the test case.
In non-testcase code path, `_userTaskManager` will be set. Hypothetically, even if we don't have `UserTaskInfo`, sending out a notification will at least alert that `Executor` state has changed. The notification will contain UUID of the UserTask, which can be used to look up request (within retention). The notification will also tell when and by what is the execution stopped. These provides more information than panic'ing as suggested, which is why we decided to keep it as is.